packages/buidler-core/src/builtin-tasks/compile.ts

```diff
@@ -250,6 +244,24 @@ async function removeObsoleteArtifacts(
   }
 }
 
+async function removeObsoleteBuildInfos(artifactsPath: string) {
+  const dbgFiles = await glob(path.join(artifactsPath, "**/*.dbg"));
+
+  const validBuildInfos = new Set<string>();
+  for (const dbgFile of dbgFiles) {
+    const { buildInfo } = await fsExtra.readJson(dbgFile);
+    validBuildInfos.add(path.resolve(path.dirname(dbgFile), buildInfo));
+  }
+
+  const buildInfoFiles = await getBuildInfoFiles(artifactsPath);
+
+  for (const buildInfoFile of buildInfoFiles) {
+    if (!validBuildInfos.has(buildInfoFile)) {
+      await fsExtra.unlink(buildInfoFile);
```

Comment:
I feel very uncomfortable doing `unlink`s of dynamic paths. What do you think of having some sort of wrapper that, before removing something, checks that the file is under the project root? Basically to avoid some weird scenario were we end up removing something by mistake outside the project (unlikely, but it would be bad if it happened).